sw/host/opentitanlib/src/rom_patch/mod.rs

```diff
@@ -0,0 +1,691 @@
+}
+
+fn deserialize_hex_usize<'de, D>(deserializer: D) -> Result<usize, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    usize::from_str_radix(s.trim_start_matches("0x"), 16).map_err(serde::de::Error::custom)
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchSubRoutineHjson {
+    #[serde(deserialize_with = "deserialize_hex_usize")]
+    offset: usize,
+    code: Vec<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchRegionHjson {
+    #[serde(deserialize_with = "deserialize_hex_u32")]
+    match_addr: u32,
+    locked: bool,
+    patch_code: Vec<String>,
+    sub_routines: Option<Vec<RomPatchSubRoutineHjson>>,
+}
+
+#[derive(Deserialize, Debug)]
+struct RomPatchHjson {
+    revision: u8,
+    #[serde(deserialize_with = "deserialize_hex_u32")]
+    base_addr: u32,
+    regions: Vec<RomPatchRegionHjson>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct RomPatchPartitionHjson {
+    patches: Vec<RomPatchHjson>,
+}
+
+#[derive(AsBytes, Default, Clone, Copy, Debug)]
+#[repr(packed, C)]
+struct RomPatchHeader {
+    meta: u8,
+    size: u16,
+    revision: u8,
+}
+
+#[derive(AsBytes, Copy, Clone, Debug, Default)]
+#[repr(packed, C)]
+struct RomPatchMatchDescriptor {
+    match_base: u32,
+    remap_base: u32,
+}
+
+// Patch locked field position in the match descriptor.
+const PATCH_LOCKED_FIELD_SHIFT: usize = 31;
+
+// Patch size field position in the match descriptor.
+const PATCH_SIZE_FIELD_SHIFT: usize = 27;
+const PATCH_SIZE_FIELD_WIDTH: usize = 4;
+const PATCH_SIZE_FIELD_MASK: u32 = (1 << PATCH_SIZE_FIELD_WIDTH) - 1;
+
+// Patch match address in the match descriptor.
+const PATCH_MATCH_ADDR_FIELD_WIDTH: usize = 27;
+const PATCH_MATCH_ADDR_FIELD_MASK: u32 = (1 << PATCH_MATCH_ADDR_FIELD_WIDTH) - 1;
+
+impl RomPatchMatchDescriptor {
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = region.match_addr;
+        let patch_code_size_words = region.patch_code.len();
+
+        let size_dwords = match patch_code_size_words {
+            1 | 2 | 4 | 8 => (patch_code_size_words as u32) & PATCH_SIZE_FIELD_MASK,
+            _ => bail!("Invalid ROM patch region code size {patch_code_size_words}"),
+        };
+        let locked = if region.locked {
+            1 << PATCH_LOCKED_FIELD_SHIFT
+        } else {
+            0
+        };
+
+        let match_address = match match_addr {
+            addr if addr <= PATCH_MATCH_ADDR_FIELD_MASK => addr & PATCH_MATCH_ADDR_FIELD_MASK,
+            _ => bail!("Invalid ROM patch match address {match_addr}"),
+        };
+
+        let match_base = locked | (size_dwords << PATCH_SIZE_FIELD_SHIFT) | match_address;
+
+        Ok(RomPatchMatchDescriptor {
+            match_base,
+            remap_base,
+        })
+    }
+}
+
+#[derive(AsBytes, Copy, Clone, Debug, Default)]
+#[repr(packed, C)]
+struct RomPatchMatchTable {
+    table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS],
```

Comment:
N_MATCH_DESCRIPTORS is an RTL constant for Darjeeling. But I guess using const generics make it more flexible. I switched the implementation to it now, and although rom_patch.rs uses the exported N_MATCH_DESCRIPTORS, this could become a CLI option in the future.